kv/kvserver: TestAtomicReplicationChange failed

kv/kvserver.TestAtomicReplicationChange [failed](https://tanzanite.cluster.engflow.com/invocations/default/76d3f2c6-ba8d-41f8-b180-a86a658e8776?testReportRun=6&testReportShard=11&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on release-25.2 @ [686d7076949cba431d34512ab1e2c15012d358b8](https://github.com/cockroachdb/cockroach/commits/686d7076949cba431d34512ab1e2c15012d358b8):


```
=== RUN   TestAtomicReplicationChange
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestAtomicReplicationChange368036001
    test_log_scope.go:76: use -show-logs to present logs inline
    client_atomic_membership_change_test.go:65: condition failed to evaluate within 45s: from client_atomic_membership_change_test.go:75: [n1,s1,r75/1:/{Table/Max-Max}] should have been removed
    panic.go:635: -- test log scope end --
test logs left over in: outputs.zip/logTestAtomicReplicationChange368036001
--- FAIL: TestAtomicReplicationChange (46.79s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=6</code>
 - <code>shard=11</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAtomicReplicationChange.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-50295

--- a/pkg/kv/kvserver/client_atomic_membership_change_test.go
+++ b/pkg/kv/kvserver/client_atomic_membership_change_test.go
@@ -8,16 +8,19 @@ package kvserver_test
 import (
 	"context"
 	"strings"
+	"sync/atomic"
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver"
+	"github.com/cockroachdb/cockroach/pkg/raft"
 	"github.com/cockroachdb/cockroach/pkg/raft/confchange"
 	"github.com/cockroachdb/cockroach/pkg/raft/quorum"
 	"github.com/cockroachdb/cockroach/pkg/raft/tracker"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -34,10 +37,16 @@ func TestAtomicReplicationChange(t *testing.T) {
 	defer log.Scope(t).Close(t)
 	ctx := context.Background()
 
+	var logRangeID atomic.Int64
 	args := base.TestClusterArgs{
 		ServerArgs: base.TestServerArgs{
 			Knobs: base.TestingKnobs{
-				Store: &kvserver.StoreTestingKnobs{},
+				Store: &kvserver.StoreTestingKnobs{
+					RaftLogReadyRaftMuLocked: func(
+						ctx context.Context, rangeID roachpb.RangeID, replID roachpb.ReplicaID, rd raft.Ready) bool {
+						return logRangeID.Load() == int64(rangeID)
+					},
+				},
 			},
 		},
 		ReplicationMode: base.ReplicationManual,
@@ -111,6 +120,16 @@ func TestAtomicReplicationChange(t *testing.T) {
 
 	// Replicas should now live on all stores except s3.
 	checkDesc(desc, 1, 2, 4, 5, 6)
+
+	// Start logging all raft Ready handling to diagnose issues such as:
+	//
+	// https://github.com/cockroachdb/cockroach/issues/145946
+	//
+	// We don't do this under Duress since this logging is fairly
+	// verbose and thus expensive.
+	if !skip.Duress() {
+		logRangeID.Store(int64(desc.RangeID))
+	}
 
 	// Transfer the lease to s5.
 	require.NoError(t, tc.TransferRangeLease(desc, tc.Target(4)))

--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
@@ -69,6 +69,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/tracing/tracingpb"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
 	"github.com/cockroachdb/errors"
+	"github.com/cockroachdb/redact"
 	"github.com/stretchr/testify/require"
 	"google.golang.org/grpc"
 )
@@ -3084,7 +3085,7 @@ func TestReplicaRemovalCampaign(t *testing.T) {
 
 			if td.remove {
 				// Simulate second replica being transferred by removing it.
-				if err := store0.RemoveReplica(ctx, replica2, replica2.Desc().NextReplicaID, kvserver.RemoveOptions{
+				if err := store0.RemoveReplica(ctx, replica2, replica2.Desc().NextReplicaID, redact.SafeString(t.Name()), kvserver.RemoveOptions{
 					DestroyData: true,
 				}); err != nil {
 					t.Fatal(err)

--- a/pkg/kv/kvserver/mvcc_gc_queue_test.go
+++ b/pkg/kv/kvserver/mvcc_gc_queue_test.go
@@ -36,6 +36,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
 	"github.com/cockroachdb/errors"
+	"github.com/cockroachdb/redact"
 	"github.com/kr/pretty"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
@@ -1510,7 +1511,7 @@ func TestMVCCGCQueueGroupsRangeDeletions(t *testing.T) {
 	store := createTestStoreWithConfig(ctx, t, stopper, testStoreOpts{createSystemRanges: false}, &cfg)
 	r, err := store.GetReplica(roachpb.RangeID(1))
 	require.NoError(t, err)
-	require.NoError(t, store.RemoveReplica(ctx, r, r.Desc().NextReplicaID, RemoveOptions{DestroyData: true}))
+	require.NoError(t, store.RemoveReplica(ctx, r, r.Desc().NextReplicaID, redact.SafeString(t.Name()), RemoveOptions{DestroyData: true}))
 	// Add replica without hint.
 	r1 := createReplica(store, roachpb.RangeID(100), key("a"), key("b"))
 	require.NoError(t, store.AddReplica(r1))

--- a/pkg/kv/kvserver/queue_test.go
+++ b/pkg/kv/kvserver/queue_test.go
@@ -33,6 +33,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/errors"
+	"github.com/cockroachdb/redact"
 	"github.com/gogo/protobuf/proto"
 	"github.com/stretchr/testify/require"
 )
@@ -122,7 +123,7 @@ func createReplicas(t *testing.T, tc *testContext, num int) []*Replica {
 	if err != nil {
 		t.Fatal(err)
 	}
-	if err := tc.store.RemoveReplica(context.Background(), repl1, repl1.Desc().NextReplicaID, RemoveOptions{
+	if err := tc.store.RemoveReplica(context.Background(), repl1, repl1.Desc().NextReplicaID, redact.SafeString(t.Name()), RemoveOptions{
 		DestroyData: true,
 	}); err != nil {
 		t.Fatal(err)
@@ -707,7 +708,7 @@ func TestAcceptsUnsplitRanges(t *testing.T) {
 	// Remove replica for range 1 since it encompasses the entire keyspace.
 	repl1, err := s.GetReplica(1)
 	require.NoError(t, err)
-	require.NoError(t, s.RemoveReplica(ctx, repl1, repl1.Desc().NextReplicaID, RemoveOptions{DestroyData: true}))
+	require.NoError(t, s.RemoveReplica(ctx, repl1, repl1.Desc().NextReplicaID, redact.SafeString(t.Name()), RemoveOptions{DestroyData: true}))
 
 	// This range can never be split due to zone configs boundaries.
 	neverSplits := createReplica(s, 2, roachpb.RKeyMin, maxWontSplitAddr)
@@ -909,7 +910,7 @@ func TestBaseQueuePurgatory(t *testing.T) {
 	// the replica set. The number of processed replicas will be 2 less.
 	const rmReplCount = 2
 	repls[0].replicaID = 2
-	if err := tc.store.RemoveReplica(ctx, repls[1], repls[1].Desc().NextReplicaID, RemoveOptions{
+	if err := tc.store.RemoveReplica(ctx, repls[1], repls[1].Desc().NextReplicaID, redact.SafeString(t.Name()), RemoveOptions{
 		DestroyData: true,
 	}); err != nil {
 		t.Fatal(err)

--- a/pkg/kv/kvserver/raft.go
+++ b/pkg/kv/kvserver/raft.go
@@ -150,11 +150,18 @@ func verboseRaftLoggingEnabled() bool {
 	return log.V(5)
 }
 
-func logRaftReady(ctx context.Context, ready raft.Ready) {
-	if !verboseRaftLoggingEnabled() {
+func (r *Replica) maybeLogRaftReadyRaftMuLocked(ctx context.Context, ready raft.Ready) {
+	fn := r.store.TestingKnobs().RaftLogReadyRaftMuLocked
+	switch {
+	case verboseRaftLoggingEnabled():
+	case fn != nil && fn(ctx, r.RangeID, r.ReplicaID(), ready):
+	default:
 		return
 	}
+	logRaftReady(ctx, ready)
+}
 
+func logRaftReady(ctx context.Context, ready raft.Ready) {
 	var buf bytes.Buffer
 	if ready.SoftState != nil {
 		fmt.Fprintf(&buf, "  SoftState updated: %+v\n", *ready.SoftState)

--- a/pkg/kv/kvserver/replica_application_result.go
+++ b/pkg/kv/kvserver/replica_application_result.go
@@ -615,10 +615,12 @@ func (r *Replica) handleChangeReplicasResult(
 		r.store.metrics.RangeRaftLeaderRemovals.Inc(1)
 	}
 
-	if _, err := r.store.removeInitializedReplicaRaftMuLocked(ctx, r, chng.NextReplicaID(), RemoveOptions{
-		// We destroyed the data when the batch committed so don't destroy it again.
-		DestroyData: false,
-	}); err != nil {
+	if _, err := r.store.removeInitializedReplicaRaftMuLocked(
+		ctx, r, chng.NextReplicaID(), "applied self-removal",
+		RemoveOptions{
+			// We destroyed the data when the batch committed so don't destroy it again.
+			DestroyData: false,
+		}); err != nil {
 		log.Fatalf(ctx, "failed to remove replica: %v", err)
 	}
 

--- a/pkg/kv/kvserver/replica_gc_queue.go
+++ b/pkg/kv/kvserver/replica_gc_queue.go
@@ -315,7 +315,7 @@ func (rgcq *replicaGCQueue) process(
 		// possible if we currently think we're processing a pre-emptive snapshot
 		// but discover in RemoveReplica that this range has since been added and
 		// knows that.
-		if err := repl.store.RemoveReplica(ctx, repl, nextReplicaID, RemoveOptions{
+		if err := repl.store.RemoveReplica(ctx, repl, nextReplicaID, "MVCC GC queue", RemoveOptions{
 			DestroyData: true,
 		}); err != nil {
 			// Should never get an error from RemoveReplica.
@@ -359,7 +359,7 @@ func (rgcq *replicaGCQueue) process(
 		// A tombstone is written with a value of mergedTombstoneReplicaID because
 		// we know the range to have been merged. See the Merge case of
 		// runPreApplyTriggers() for details.
-		if err := repl.store.RemoveReplica(ctx, repl, mergedTombstoneReplicaID, RemoveOptions{
+		if err := repl.store.RemoveReplica(ctx, repl, mergedTombstoneReplicaID, "dangling subsume via MVCC GC queue", RemoveOptions{
 			DestroyData: true,
 		}); err != nil {
 			return false, err

--- a/pkg/kv/kvserver/replica_raft.go
+++ b/pkg/kv/kvserver/replica_raft.go
@@ -1015,7 +1015,7 @@ func (r *Replica) handleRaftReadyRaftMuLocked(
 	}
 
 	if hasReady {
-		logRaftReady(ctx, ready)
+		r.maybeLogRaftReadyRaftMuLocked(ctx, ready)
 	}
 	// Even if we don't have a Ready, or entries in Ready,
 	// replica_rac2.Processor may need to do some work.

--- a/pkg/kv/kvserver/replica_raftstorage.go
+++ b/pkg/kv/kvserver/replica_raftstorage.go
@@ -502,7 +502,7 @@ func (r *Replica) applySnapshotRaftMuLocked(
 		logDetails.Printf(" excise=true")
 		logDetails.Printf(" ingestion=%d@%0.0fms", len(inSnap.SSTStorageScratch.SSTs()),
 			stats.ingestion.Sub(stats.subsumedReplicas).Seconds()*1000)
-		var appliedAsWriteStr string
+		var appliedAsWriteStr redact.SafeString
 		if !applyAsIngest {
 			appliedAsWriteStr = "as write "
 		}
@@ -780,11 +780,13 @@ func (r *Replica) clearSubsumedReplicaInMemoryData(
 		// consume. Without this, there would be a risk of errant snapshots being
 		// allowed in (perhaps not involving any of the RangeIDs known to the merge
 		// but still touching its keyspace) and causing corruption.
-		ph, err := r.store.removeInitializedReplicaRaftMuLocked(ctx, sr, subsumedNextReplicaID, RemoveOptions{
-			// The data was already destroyed by clearSubsumedReplicaDiskData.
-			DestroyData:       false,
-			InsertPlaceholder: true,
-		})
+		ph, err := r.store.removeInitializedReplicaRaftMuLocked(
+			ctx, sr, subsumedNextReplicaID, "subsumed by snapshot",
+			RemoveOptions{
+				// The data was already destroyed by clearSubsumedReplicaDiskData.
+				DestroyData:       false,
+				InsertPlaceholder: true,
+			})
 		if err != nil {
 			return nil, err
 		}

--- a/pkg/kv/kvserver/replica_test.go
+++ b/pkg/kv/kvserver/replica_test.go
@@ -7190,9 +7190,11 @@ func TestReplicaDestroy(t *testing.T) {
 	func() {
 		tc.repl.raftMu.Lock()
 		defer tc.repl.raftMu.Unlock()
-		_, err := tc.store.removeInitializedReplicaRaftMuLocked(ctx, tc.repl, repl.Desc().NextReplicaID, RemoveOptions{
-			DestroyData: true,
-		})
+		_, err := tc.store.removeInitializedReplicaRaftMuLocked(
+			ctx, tc.repl, repl.Desc().NextReplicaID, redact.SafeString(t.Name()),
+			RemoveOptions{
+				DestroyData: true,
+			})
 		require.NoError(t, err)
 	}()
 
@@ -7372,9 +7374,11 @@ func TestQuotaPoolAccessOnDestroyedReplica(t *testing.T) {
 	func() {
 		tc.repl.raftMu.Lock()
 		defer tc.repl.raftMu.Unlock()
-		if _, err := tc.store.removeInitializedReplicaRaftMuLocked(ctx, repl, repl.Desc().NextReplicaID, RemoveOptions{
-			DestroyData: true,
-		}); err != nil {
+		if _, err := tc.store.removeInitializedReplicaRaftMuLocked(
+			ctx, repl, repl.Desc().NextReplicaID, redact.SafeString(t.Name()),
+			RemoveOptions{
+				DestroyData: true,
+			}); err != nil {
 			t.Fatal(err)
 		}
 	}()
@@ -9411,8 +9415,7 @@ func TestCancelPendingCommands(t *testing.T) {
 		t.Fatalf("command finished earlier than expected with error %v", pErr)
 	default:
 	}
-	require.NoError(t, tc.store.RemoveReplica(ctx, tc.repl, tc.repl.Desc().NextReplicaID,
-		RemoveOptions{DestroyData: true}))
+	require.NoError(t, tc.store.RemoveReplica(ctx, tc.repl, tc.repl.Desc().NextReplicaID, redact.SafeString(t.Name()), RemoveOptions{DestroyData: true}))
 	pErr := <-errChan
 	if _, ok := pErr.GetDetail().(*kvpb.AmbiguousResultError); !ok {
 		t.Errorf("expected AmbiguousResultError, got %v", pErr)

--- a/pkg/kv/kvserver/store_create_replica.go
+++ b/pkg/kv/kvserver/store_create_replica.go
@@ -127,7 +127,7 @@ func (s *Store) tryGetReplica(
 		}
 
 		repl.mu.RUnlock()
-		if err := s.removeReplicaRaftMuLocked(ctx, repl, replicaID, RemoveOptions{
+		if err := s.removeReplicaRaftMuLocked(ctx, repl, replicaID, "superseded by newer Replica", RemoveOptions{
 			DestroyData: true,
 		}); err != nil {
 			log.Fatalf(ctx, "failed to remove replica: %v", err)

--- a/pkg/kv/kvserver/store_merge.go
+++ b/pkg/kv/kvserver/store_merge.go
@@ -126,12 +126,14 @@ func (s *Store) MergeRange(
 	//
 	// We ask removeInitializedReplicaRaftMuLocked to install a placeholder which
 	// we'll drop atomically with extending the right-hand side down below.
-	ph, err := s.removeInitializedReplicaRaftMuLocked(ctx, rightRepl, rightDesc.NextReplicaID, RemoveOptions{
-		// The replica was destroyed by the tombstones added to the batch in
-		// runPostAddTriggers.
-		DestroyData:       false,
-		InsertPlaceholder: true,
-	})
+	ph, err := s.removeInitializedReplicaRaftMuLocked(
+		ctx, rightRepl, rightDesc.NextReplicaID, "merge trigger",
+		RemoveOptions{
+			// The replica was destroyed by the tombstones added to the batch in
+			// runPostAddTriggers.
+			DestroyData:       false,
+			InsertPlaceholder: true,
+		})
 	if err != nil {
 		return errors.Wrap(err, "cannot remove range")
 	}

--- a/pkg/kv/kvserver/store_raft.go
+++ b/pkg/kv/kvserver/store_raft.go
@@ -641,7 +641,7 @@ func (s *Store) HandleRaftResponse(
 
 				repl.mu.Unlock()
 				nextReplicaID := tErr.ReplicaID + 1
-				return s.removeReplicaRaftMuLocked(ctx, repl, nextReplicaID, RemoveOptions{
+				return s.removeReplicaRaftMuLocked(ctx, repl, nextReplicaID, "received ReplicaTooOldError", RemoveOptions{
 					DestroyData: true,
 				})
 			case *kvpb.RaftGroupDeletedError:

--- a/pkg/kv/kvserver/store_remove_replica.go
+++ b/pkg/kv/kvserver/store_remove_replica.go
@@ -13,6 +13,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/errors"
+	"github.com/cockroachdb/redact"
 )
 
 // RemoveOptions bundles boolean parameters for Store.RemoveReplica.
@@ -36,28 +37,36 @@ type RemoveOptions struct {
 //
 // The passed replica must be initialized.
 func (s *Store) RemoveReplica(
-	ctx context.Context, rep *Replica, nextReplicaID roachpb.ReplicaID, opts RemoveOptions,
+	ctx context.Context,
+	rep *Replica,
+	nextReplicaID roachpb.ReplicaID,
+	reason redact.SafeString,
+	opts RemoveOptions,
 ) error {
 	rep.raftMu.Lock()
 	defer rep.raftMu.Unlock()
 	if opts.InsertPlaceholder {
 		return errors.Errorf("InsertPlaceholder not supported in RemoveReplica")
 	}
-	_, err := s.removeInitializedReplicaRaftMuLocked(ctx, rep, nextReplicaID, opts)
+	_, err := s.removeInitializedReplicaRaftMuLocked(ctx, rep, nextReplicaID, reason, opts)
 	return err
 }
 
 // removeReplicaRaftMuLocked removes the passed replica. If the replica is
 // initialized the RemoveOptions will be consulted.
 func (s *Store) removeReplicaRaftMuLocked(
-	ctx context.Context, rep *Replica, nextReplicaID roachpb.ReplicaID, opts RemoveOptions,
+	ctx context.Context,
+	rep *Replica,
+	nextReplicaID roachpb.ReplicaID,
+	reason redact.SafeString,
+	opts RemoveOptions,
 ) error {
 	rep.raftMu.AssertHeld()
 	if rep.IsInitialized() {
 		if opts.InsertPlaceholder {
 			return errors.Errorf("InsertPlaceholder unsupported in removeReplicaRaftMuLocked")
 		}
-		_, err := s.removeInitializedReplicaRaftMuLocked(ctx, rep, nextReplicaID, opts)
+		_, err := s.removeInitializedReplicaRaftMuLocked(ctx, rep, nextReplicaID, reason, opts)
 		return errors.Wrap(err,
 			"failed to remove replica")
 	}
@@ -69,7 +78,11 @@ func (s *Store) removeReplicaRaftMuLocked(
 // which is sometimes called directly when the necessary lock is already held.
 // It requires that Replica.raftMu is held and that s.mu is not held.
 func (s *Store) removeInitializedReplicaRaftMuLocked(
-	ctx context.Context, rep *Replica, nextReplicaID roachpb.ReplicaID, opts RemoveOptions,
+	ctx context.Context,
+	rep *Replica,
+	nextReplicaID roachpb.ReplicaID,
+	reason redact.SafeString,
+	opts RemoveOptions,
 ) (*ReplicaPlaceholder, error) {
 	rep.raftMu.AssertHeld()
 	if !rep.IsInitialized() {
@@ -149,7 +162,7 @@ func (s *Store) removeInitializedReplicaRaftMuLocked(
 
 	// During merges, the context might have the subsuming range, so we explicitly
 	// log the replica to be removed.
-	log.Infof(ctx, "removing replica r%d/%d", rep.RangeID, rep.replicaID)
+	log.Infof(ctx, "removing replica r%d/%d (%s)", rep.RangeID, rep.replicaID, reason)
 
 	s.mu.Lock()
 	if it := s.getOverlappingKeyRangeLocked(desc); it.repl != rep {

--- a/pkg/kv/kvserver/store_test.go
+++ b/pkg/kv/kvserver/store_test.go
@@ -71,6 +71,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
 	"github.com/cockroachdb/errors"
+	"github.com/cockroachdb/redact"
 	"github.com/kr/pretty"
 	"github.com/stretchr/testify/require"
 	"golang.org/x/sync/errgroup"
@@ -623,7 +624,7 @@ func TestStoreAddRemoveRanges(t *testing.T) {
 		t.Error(err)
 	}
 	// Remove range 1.
-	if err := store.RemoveReplica(ctx, repl1, repl1.Desc().NextReplicaID, RemoveOptions{
+	if err := store.RemoveReplica(ctx, repl1, repl1.Desc().NextReplicaID, redact.SafeString(t.Name()), RemoveOptions{
 		DestroyData: true,
 	}); err != nil {
 		t.Error(err)
@@ -639,7 +640,7 @@ func TestStoreAddRemoveRanges(t *testing.T) {
 		t.Fatal("expected error re-adding same range")
 	}
 	// Try to remove range 1 again.
-	if err := store.RemoveReplica(ctx, repl1, repl1.Desc().NextReplicaID, RemoveOptions{
+	if err := store.RemoveReplica(ctx, repl1, repl1.Desc().NextReplicaID, redact.SafeString(t.Name()), RemoveOptions{
 		DestroyData: true,
 	}); err != nil {
 		t.Fatalf("didn't expect error re-removing same range: %v", err)
@@ -747,7 +748,7 @@ func TestStoreRemoveReplicaDestroy(t *testing.T) {
 
 	// Can't remove Replica with DestroyData false because this requires the destroyStatus
 	// to already have been set by the caller (but we didn't).
-	require.ErrorContains(t, store.RemoveReplica(ctx, repl1, repl1.Desc().NextReplicaID, RemoveOptions{
+	require.ErrorContains(t, store.RemoveReplica(ctx, repl1, repl1.Desc().NextReplicaID, redact.SafeString(t.Name()), RemoveOptions{
 		DestroyData: false,
 	}), `replica not marked as destroyed`)
 
@@ -755,14 +756,14 @@ func TestStoreRemoveReplicaDestroy(t *testing.T) {
 	// NB: we rely on this idempotency today (as @tbg found out when he accidentally
 	// removed it).
 	for i := 0; i < 2; i++ {
-		require.NoError(t, store.RemoveReplica(ctx, repl1, repl1.Desc().NextReplicaID, RemoveOptions{
+		require.NoError(t, store.RemoveReplica(ctx, repl1, repl1.Desc().NextReplicaID, redact.SafeString(t.Name()), RemoveOptions{
 			DestroyData: true,
 		}), "%d", i)
 	}
 
 	// However, if we have DestroyData=false, caller is expected to be the unique first "destroyer"
 	// of the Replica.
-	require.ErrorContains(t, store.RemoveReplica(ctx, repl1, repl1.Desc().NextReplicaID, RemoveOptions{
+	require.ErrorContains(t, store.RemoveReplica(ctx, repl1, repl1.Desc().NextReplicaID, redact.SafeString(t.Name()), RemoveOptions{
 		DestroyData: false,
 	}), `does not exist`)
 
@@ -805,7 +806,7 @@ func TestStoreReplicaVisitor(t *testing.T) {
 	if err != nil {
 		t.Error(err)
 	}
-	if err := store.RemoveReplica(ctx, repl1, repl1.Desc().NextReplicaID, RemoveOptions{
+	if err := store.RemoveReplica(ctx, repl1, repl1.Desc().NextReplicaID, redact.SafeString(t.Name()), RemoveOptions{
 		DestroyData: true,
 	}); err != nil {
 		t.Error(err)
@@ -889,7 +890,7 @@ func TestMarkReplicaInitialized(t *testing.T) {
 	if err != nil {
 		t.Error(err)
 	}
-	if err := store.RemoveReplica(ctx, repl1, repl1.Desc().NextReplicaID, RemoveOptions{
+	if err := store.RemoveReplica(ctx, repl1, repl1.Desc().NextReplicaID, redact.SafeString(t.Name()), RemoveOptions{
 		DestroyData: true,
 	}); err != nil {
 		t.Error(err)
@@ -2761,7 +2762,7 @@ func TestMaybeRemove(t *testing.T) {
 	if err != nil {
 		t.Error(err)
 	}
-	if err := store.RemoveReplica(ctx, repl, repl.Desc().NextReplicaID, RemoveOptions{
+	if err := store.RemoveReplica(ctx, repl, repl.Desc().NextReplicaID, redact.SafeString(t.Name()), RemoveOptions{
 		DestroyData: true,
 	}); err != nil {
 		t.Error(err)
@@ -2878,7 +2879,7 @@ func TestStoreRangePlaceholders(t *testing.T) {
 	if err != nil {
 		t.Error(err)
 	}
-	if err := s.RemoveReplica(ctx, repl1, repl1.Desc().NextReplicaID, RemoveOptions{
+	if err := s.RemoveReplica(ctx, repl1, repl1.Desc().NextReplicaID, redact.SafeString(t.Name()), RemoveOptions{
 		DestroyData: true,
 	}); err != nil {
 		t.Error(err)
@@ -3014,7 +3015,7 @@ func TestStoreRemovePlaceholderOnRaftIgnored(t *testing.T) {
 	repl1, err := s.GetReplica(1)
 	desc := repl1.Desc()
 	require.NoError(t, err)
-	require.NoError(t, s.RemoveReplica(ctx, repl1, desc.NextReplicaID, RemoveOptions{
+	require.NoError(t, s.RemoveReplica(ctx, repl1, desc.NextReplicaID, redact.SafeString(t.Name()), RemoveOptions{
 		DestroyData: true,
 	}))
 

--- a/pkg/kv/kvserver/testing_knobs.go
+++ b/pkg/kv/kvserver/testing_knobs.go
@@ -37,17 +37,23 @@ import (
 // particular point is reached) or to change the behavior by returning
 // an error (which aborts all further processing for the command).
 type StoreTestingKnobs struct {
-	EvalKnobs               kvserverbase.BatchEvalTestingKnobs
-	IntentResolverKnobs     kvserverbase.IntentResolverTestingKnobs
-	TxnWaitKnobs            txnwait.TestingKnobs
-	ConsistencyTestingKnobs ConsistencyTestingKnobs
-	TenantRateKnobs         tenantrate.TestingKnobs
-	EngineKnobs             []storage.ConfigOption
-	AllocatorKnobs          *allocator.TestingKnobs
-	GossipTestingKnobs      StoreGossipTestingKnobs
-	ReplicaPlannerKnobs     plan.ReplicaPlannerTestingKnobs
-	StoreLivenessKnobs      *storeliveness.TestingKnobs
-	RaftTestingKnobs        *raft.TestingKnobs
+	EvalKnobs                kvserverbase.BatchEvalTestingKnobs
+	IntentResolverKnobs      kvserverbase.IntentResolverTestingKnobs
+	TxnWaitKnobs             txnwait.TestingKnobs
+	ConsistencyTestingKnobs  ConsistencyTestingKnobs
+	TenantRateKnobs          tenantrate.TestingKnobs
+	EngineKnobs              []storage.ConfigOption
+	AllocatorKnobs           *allocator.TestingKnobs
+	GossipTestingKnobs       StoreGossipTestingKnobs
+	ReplicaPlannerKnobs      plan.ReplicaPlannerTestingKnobs
+	StoreLivenessKnobs       *storeliveness.TestingKnobs
+	RaftTestingKnobs         *raft.TestingKnobs
+	RaftLogReadyRaftMuLocked func(
+		ctx context.Context,
+		rangeID roachpb.RangeID,
+		replID roachpb.ReplicaID,
+		rd raft.Ready,
+	) bool
 	// TestingRequestFilter is called before evaluating each request on a
 	// replica. The filter is run before the request acquires latches, so
 	// blocking in the filter will not block interfering requests. If it

--- a/pkg/server/structlogging/hot_ranges_log.go
+++ b/pkg/server/structlogging/hot_ranges_log.go
@@ -117,9 +117,7 @@ func StartHotRangesLoggingScheduler(
 // installation.
 func (s *hotRangesLoggingScheduler) startTask(ctx context.Context, stopper *stop.Stopper) error {
 	return stopper.RunAsyncTask(ctx, "hot-ranges-stats", func(ctx context.Context) {
-		if err := s.start(ctx, stopper); err != nil {
-			log.Warningf(ctx, "hot ranges stats logging scheduler stopped: %v", err)
-		}
+		s.start(ctx, stopper)
 	})
 }
 
@@ -134,7 +132,7 @@ func (s *hotRangesLoggingScheduler) startJob() error {
 	return nil
 }
 
-func (s *hotRangesLoggingScheduler) start(ctx context.Context, stopper *stop.Stopper) error {
+func (s *hotRangesLoggingScheduler) start(ctx context.Context, stopper *stop.Stopper) {
 	for {
 		ci := CheckInterval
 		if s.multiTenant {
@@ -142,9 +140,9 @@ func (s *hotRangesLoggingScheduler) start(ctx context.Context, stopper *stop.Sto
 		}
 		select {
 		case <-stopper.ShouldQuiesce():
-			return nil
+			return
 		case <-ctx.Done():
-			return nil
+			return
 		case <-time.After(ci):
 			s.maybeLogHotRanges(ctx, stopper)
 		case <-TestLoopChannel:

--- a/pkg/server/structlogging/hot_ranges_log.go
+++ b/pkg/server/structlogging/hot_ranges_log.go
@@ -117,8 +117,9 @@ func StartHotRangesLoggingScheduler(
 // installation.
 func (s *hotRangesLoggingScheduler) startTask(ctx context.Context, stopper *stop.Stopper) error {
 	return stopper.RunAsyncTask(ctx, "hot-ranges-stats", func(ctx context.Context) {
-		err := s.start(ctx, stopper)
-		log.Warningf(ctx, "hot ranges stats logging scheduler stopped: %s", err)
+		if err := s.start(ctx, stopper); err != nil {
+			log.Warningf(ctx, "hot ranges stats logging scheduler stopped: %v", err)
+		}
 	})
 }
 

--- a/pkg/server/structlogging/hot_ranges_log_job.go
+++ b/pkg/server/structlogging/hot_ranges_log_job.go
@@ -26,7 +26,8 @@ func (s *hotRangesLoggingScheduler) Resume(ctx context.Context, execCtxI interfa
 	// terminate the SQL pod whenever the job is running, so mark it as idle.
 	s.job.MarkIdle(true)
 
-	return s.start(ctx, s.stopper)
+	s.start(ctx, s.stopper)
+	return nil
 }
 
 func (s *hotRangesLoggingScheduler) OnFailOrCancel(

